### PR TITLE
SAK-48432 SAMIGO Problem on Calculated question with e-4

### PR DESCRIPTION
**Pull request by jesusmmp**
https://sakaiproject.atlassian.net/browse/SAK-48432

There are problems with some some test... mmmm

image

Nowadays, this test code should be...

```
result = gradingService.processFormulaIntoValue("1.8E-4", 2);
//Assert.assertEquals("1.8E-4", result);
Assert.assertEquals("0", result); 
```

I think is incorrect...

### CODE REVIEW

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java [hunk 1]
```diff
@@ -2952,7 +2952,11 @@ public String toScientificNotation(String numberStr,int decimalPlaces){
 	    && bdx.doubleValue() != 0) {
 		  formatter = new DecimalFormat(FORMAT_MASK);
 	  } else {
-		  formatter = new DecimalFormat("0");
+		  if ((String.valueOf(bdx.doubleValue()).contains("e")) || ((String.valueOf(bdx.doubleValue()).contains("E")))) {
```
Comment (bgarciaentornos) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java:2955]: Quite minor, but maybe we could use StringUtils.containsIgnoreCase or StringUtils.containsAny

**Comment by ottenhoff** [comment]
I don't really understand what you're attempting to do here.

I don't think most instructors or students want scientific notation used except with very large numbers.

**Comment by jesusmmp** [comment]
> 

Instructions

{A} {dB}  {r} 

{{P}}   

Variables

A -  0.28 - 0.28 - 2
db - 65   - 65   - 0
r  - 2.1  - 2.1  - 1

Solution

P - 1e-1210^({dB}/10)4pi{r}^2

Result should  1.8e-4 be instead of 0

image

**Comment by jonespm** [comment]
Can you add new test case or update the test case that this fixes in GradingServiceTest.java? I'm not sure what this is intending to resolve. The code should already be handling this. Is the `DEFAULT_MIN_THRESHOLD ` too high?

This looks like it's causing a test failure now so something needs  to be updated.

```
Error: [ERROR] Failures: 
Error: [ERROR]   GradingServiceTest.testCalculationsToValues:453 expected:<1[000000000]0> but was:<1[E1]0>
```

**Comment by jesusmmp** [comment]
> Can you add new test case or update the test case that this fixes in GradingServiceTest.java? I'm not sure what this is intending to resolve. The code should already be handling this. Is the `DEFAULT_MIN_THRESHOLD ` too high?
> 
> This looks like it's causing a test failure now so something needs to be updated.
> 
> ```
> Error: [ERROR] Failures: 
> Error: [ERROR]   GradingServiceTest.testCalculationsToValues:453 expected:<1[000000000]0> but was:<1[E1]0>
> ```

Hi @jonespm 

**Case 1:**
If expr = 1.8E-8

 e = new Expression(expr);
double d = e.calculate();
ans = new BigDecimal(d, MathContext.DECIMAL64);

Result:
d = 1.8E-8
ans = 1.8E-8

**Case 2:**
If expr = 1.8E-4

e = new Expression(expr);
double d = e.calculate();
ans = new BigDecimal(d, MathContext.DECIMAL64);

Result:
d = 1.8E-4
ans = 0.000175

---

If we reduced the MIN_THRESHOLD to 0.01 then both cases works but I don't know if other cases fails...

**Comment by jesusmmp** [comment]
Just updated!. All tests are OK. Added some tests related to the problem.

**Comment by jesusmmp** [comment]
> @jesusmmp this has conflicts can you update, also lets talk about this PR at the next core call

Conflicts resolved. Just updated!

**Review by ern — APPROVED** [review]

**Pull request merged**